jobs/cloud/worker.go

```diff
@@ -0,0 +1,284 @@
+		}
+	}
+
+	if forgivenessCount >= 3 {
+		// There's no forgiving someone who's broken our hearts 3 or more times
+		return false
+	}
+
+	forgivenessCount++
+	sysVar := &model.System{Name: model.OVER_USER_LIMIT_FORGIVEN_COUNT, Value: strconv.Itoa(forgivenessCount)}
+	err = worker.app.Srv().Store.System().SaveOrUpdate(sysVar)
+	if err != nil {
+		mlog.Error("Unable to save OVER_USER_LIMIT_FORGIVEN_COUNT", mlog.String("worker", worker.name), mlog.String("error", err.Error()))
+	}
+
+	_, err = worker.app.Srv().Store.System().PermanentDeleteByName(model.USER_LIMIT_OVERAGE_CYCLE_END_DATE)
+	if err != nil {
+		mlog.Error("Unable to reset USER_LIMIT_OVERAGE_CYCLE_END_DATE", mlog.String("worker", worker.name), mlog.String("error", err.Error()))
+	}
+
+	// Forgiven!
+	return true
+}
+
+func (worker *Worker) DoJob(job *model.Job) {
+	if claimed, err := worker.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", worker.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+
+	cloudUserLimit := *worker.app.Config().ExperimentalSettings.CloudUserLimit
+
+	var subscription *model.Subscription
+	var subErr *model.AppError
+	if worker.app.Srv().License() != nil && *worker.app.Srv().License().Features.Cloud {
+		subscription, subErr = worker.app.Cloud().GetSubscription()
+
+		if subErr != nil {
+			mlog.Error("Worker: Failed to retrieve subscription", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.String("error", subErr.Error()))
+			worker.setJobError(job, subErr)
+			return
+		}
+
+		if subscription == nil {
+			mlog.Error("Worker: Failed to retrieve subscription", mlog.String("worker", worker.name), mlog.String("job_id", job.Id))
+			return
+		}
+	}
+
+	if subscription != nil && subscription.IsPaidTier == "true" {
+		mlog.Info("On Paid Tier, exiting", mlog.String("worker", worker.name))
+		return
+	}
+
+	count, err := worker.app.Srv().Store.User().Count(model.UserCountOptions{IncludeDeleted: false})
+
+	if err != nil {
+		mlog.Error("Worker: Failed to get active user count", mlog.String("worker", worker.name), mlog.String("job_id", job.Id), mlog.String("error", err.Error()))
+		return
+	}
+
+	userDifference := cloudUserLimit - count
+
+	systemValue, nErr := worker.app.Srv().Store.System().GetByName(model.USER_LIMIT_OVERAGE_CYCLE_END_DATE)
+	dateLayout := "2006-01-02"
+
+	if nErr != nil {
+		mlog.Error("Error getting USER_LIMIT_OVERAGE_CYCLE_END_DATE from system store", mlog.String("worker", worker.name), mlog.String("error", nErr.Error()))
+	}
+
+	if systemValue == nil {
+		mlog.Info("Found nil end date")
+		if userDifference >= 0 {
+			// Under user limit, so no need to start tracking yet.
+			worker.LogAndSetJobSuccess(job)
+			return
+		}
+		// Our row doesn't exist, so create it and get it from the store
+		subCycleEndDate := time.Unix((subscription.EndAt / 1000), 0)
```

Comment:
Does the subscription `EndAt` value change for each monthly cycle or do we need to check the invoice?

Reply:
Yes, the EndAt is actually the `current_period_end` data from stripe, as can be seen here: https://github.com/mattermost/customer-web-server/blob/cce87a1762885e6f65eeb0db610b66c67b0b0c00/payment/stripe.go#L580

So it changes according the current cycle